### ISSUE TITLE
Subscript out of range in get_material_key

### ISSUE THREAD

**Issue by John-Nagle** [BVE-Reborn/rend3#237]
Upgraded to Rend3 0.5.0. Crash on first test.
The display comes up, and some objects are drawn, but after about 1s there is a crash.

./slscenetester --dumpfile ../../../testdata/boardroom.json 
Logging to "/tmp/log.txt"
03:41:26 [WARN] No config found!
03:41:26 [WARN] EGL says it can present to the window but not natively. This has been confirmed to malfunction on Intel+NV laptops. Therefore, we disable presentation entirely for this platform
03:41:26 [WARN] Surface does not support present mode: Mailbox, falling back to FIFO
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unknown decoration Restrict
03:41:26 [WARN] Unexpected varying type: Array { base: [4], size: Constant([8]), stride: 4 }
03:41:26 [WARN] Unexpected varying type: Array { base: [4], size: Constant([5]), stride: 4 }
Reading gltf file: /home/john/projects/sl/SL-test-viewer/slscenetester/data/unitcube2.gltf
03:41:26 [WARN] Surface does not support present mode: Mailbox, falling back to FIFO
Region offset for "Boardroom": [0, 0, 0]
Add built-in object at Vec3(128.0, 21.0, 128.0) size Vec3(256.0, 1.0, 256.0)
thread 'main' panicked at 'index out of bounds: the len is 212 but the index is 212', /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/material.rs:422:15
stack backtrace:
   0: rust_begin_unwind
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:69:5
   3: rend3::resources::material::get_material_key
   4: rend3::resources::object::ObjectManager::fill
   5: rend3::renderer::render::render_loop
   6: slscenetester::displaystate::{{closure}}
   7: winit::platform_impl::platform::x11::EventLoop<T>::run
   8: winit::platform_impl::platform::EventLoop<T>::run
   9: winit::event_loop::EventLoop<T>::run
  10: slscenetester::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
john@user-desktop:~/projects/sl/SL-test-viewer/slscenetester/target/release$

**Comment by John-Nagle** [comment]
[log.txt](https://github.com/BVE-Reborn/rend3/files/7326746/log.txt)

Full log file from above crash.

**Comment by John-Nagle** [comment]
Additional notes:

1. If I force the scene to load with no objects other than the skybox and grass ground plane, that works. Rendering and camera movement work fine.
2. The crash is occurring when some objects have loaded and have appeared on screen, but long before texture loading starts. So the objects are being drawn with a built-in grey texture shared by all objects. Nothing exciting has happened with materials yet.
3. The logged message "EGL says it can present to the window but not natively. This has been confirmed to malfunction on Intel+NV laptops. Therefore, we disable presentation entirely for this platform" indicates an Intel laptop. This is an "AMD Ryzen 5 5600X 6-Core Processor" desktop.

**Comment by CatCode79** [comment]
I found the message "EGL says it can present to the window but not natively. This has been confirmed to malfunction on Intel+NV laptops. Therefore, we disable presentation entirely for this platform" inside the wgu-hal crate.

If I understood the code the error message would be interpreted more as:
"EGL says it can present to the window but not natively. Therefore, we disable presentation entirely for this surface. (This has been confirmed to malfunction on Intel+NV laptops)

Possible connection with [Presentation disabled despite not using a laptop](https://github.com/gfx-rs/wgpu/issues/2047)

**Comment by John-Nagle** [comment]
![rend0 5 0justbeforecrash](https://user-images.githubusercontent.com/942608/137024141-9277dcfc-559d-4479-a00a-6832bcb467df.png)

Here's what displays just before the crash. The model is being loaded and being displayed properly, and drawing has been going on for over 2 seconds. But before 3 seconds, there is a crash. This is repeatable and occurs in both release and debug builds. Full backtrace below.

```
john@user-desktop:~/projects/sl/SL-test-viewer/slscenetester/target/debug$ ./slscenetester --dumpfile ../../../testdata/boardroom.json 
Logging to "/tmp/log.txt"
20:27:38 [WARN] Unable to find layer: VK_LAYER_KHRONOS_validation
20:27:38 [WARN] No config found!
20:27:38 [WARN] EGL says it can present to the window but not natively. This has been confirmed to malfunction on Intel+NV laptops. Therefore, we disable presentation entirely for this platform
20:27:38 [WARN] Surface does not support present mode: Mailbox, falling back to FIFO
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unknown decoration Restrict
20:27:38 [WARN] Unexpected varying type: Array { base: [4], size: Constant([8]), stride: 4 }
20:27:38 [WARN] Unexpected varying type: Array { base: [4], size: Constant([5]), stride: 4 }
basisu_transcoder::basisu_transcoder_init: Initializing (this is not an error)
Reading gltf file: /home/john/projects/sl/SL-test-viewer/slscenetester/data/unitcube2.gltf
Region offset for "Boardroom": [0, 0, 0]
Add built-in object at Vec3(128.0, 21.0, 128.0) size Vec3(256.0, 1.0, 256.0)
20:27:41 [WARN] Surface does not support present mode: Mailbox, falling back to FIFO
00059 frames over 01.00s. Min: 06.27ms; Average: 17.03ms; 95%: 29.56ms; 99%: 46.76ms; Max: 46.76ms; StdDev: 04.65ms
00059 frames over 01.00s. Min: 09.85ms; Average: 17.04ms; 95%: 25.00ms; 99%: 26.07ms; Max: 26.07ms; StdDev: 02.96ms
thread 'main' panicked at 'index out of bounds: the len is 212 but the index is 212', /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/material.rs:422:15
stack backtrace:
   0:     0x559a8992b940 - std::backtrace_rs::backtrace::libunwind::trace::ha0ad43e8a952bfe7
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x559a8992b940 - std::backtrace_rs::backtrace::trace_unsynchronized::h6830419c0c4130dc
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x559a8992b940 - std::sys_common::backtrace::_print_fmt::h8f3516631ffa1ef5
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x559a8992b940 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::he1640d5f0d93f618
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x559a8994f51c - core::fmt::write::h88012e1f01caeebf
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/fmt/mod.rs:1115:17
   5:     0x559a899246e5 - std::io::Write::write_fmt::h360fa85b30182555
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/io/mod.rs:1665:15
   6:     0x559a8992da5b - std::sys_common::backtrace::_print::ha1f00492f406a015
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x559a8992da5b - std::sys_common::backtrace::print::hd54561b13feb6af3
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x559a8992da5b - std::panicking::default_hook::{{closure}}::h84fe124cd0864662
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:208:50
   9:     0x559a8992d531 - std::panicking::default_hook::h5a8e74a76ce290a7
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:225:9
  10:     0x559a8992e124 - std::panicking::rust_panic_with_hook::h67c812a4fe9d4c91
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:622:17
  11:     0x559a8992dc07 - std::panicking::begin_panic_handler::{{closure}}::h33f9c1b96af300d7
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:519:13
  12:     0x559a8992be3c - std::sys_common::backtrace::__rust_end_short_backtrace::h51bae64be5921f0e
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x559a8992db69 - rust_begin_unwind
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:515:5
  14:     0x559a88180da1 - core::panicking::panic_fmt::h12a3a3c256485fca
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:92:14
  15:     0x559a88180d62 - core::panicking::panic_bounds_check::hf6f9b9870a3b97cf
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:69:5
  16:     0x559a88f7aa04 - rend3::resources::material::get_material_key::hd8537703a184b48d
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/material.rs:422:15
  17:     0x559a88fdf73b - rend3::resources::material::MaterialManager::get_material_key_and_objects::h084e2c57a2144be3
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/material.rs:306:24
  18:     0x559a88fea081 - rend3::resources::object::ObjectManager::fill::hf434d2aa4565456b
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/object.rs:67:43
  19:     0x559a88220e2e - rend3::renderer::render::render_loop::ha5aa13bb028fd96d
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/renderer/render.rs:83:21
  20:     0x559a882514aa - rend3::renderer::Renderer::render::hb811bdf2a11df2ef
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/renderer/mod.rs:391:9
  21:     0x559a8831901d - slscenetester::displaystate::{{closure}}::hc5a08615dcf04b32
                               at /home/john/projects/sl/SL-test-viewer/slscenetester/src/main.rs:683:40
  22:     0x559a88272047 - winit::platform_impl::platform::sticky_exit_callback::hb6f6cce64afedd46
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/mod.rs:746:5
  23:     0x559a882e6bf4 - winit::platform_impl::platform::x11::EventLoop<T>::run_return::hbd2a0d2ed82d048f
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/x11/mod.rs:307:21
  24:     0x559a882e7a81 - winit::platform_impl::platform::x11::EventLoop<T>::run::ha84842745d385a26
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/x11/mod.rs:385:9
  25:     0x559a88271cae - winit::platform_impl::platform::EventLoop<T>::run::h76ef47d720ce669c
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/mod.rs:662:56
  26:     0x559a8821e10a - winit::event_loop::EventLoop<T>::run::hd4d2cd9e864f8793
                               at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/event_loop.rs:154:9
  27:     0x559a882276f6 - slscenetester::displaystate::h5541896f4aced170
                               at /home/john/projects/sl/SL-test-viewer/slscenetester/src/main.rs:512:5
  28:     0x559a88227d7b - slscenetester::main::h7fc744140d3810f7
                               at /home/john/projects/sl/SL-test-viewer/slscenetester/src/main.rs:799:5
  29:     0x559a882fe13b - core::ops::function::FnOnce::call_once::hd7628fa1d8e7a697
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/ops/function.rs:227:5
  30:     0x559a8819f6ce - std::sys_common::backtrace::__rust_begin_short_backtrace::h2d0df5f0b5c25fae
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/sys_common/backtrace.rs:125:18
  31:     0x559a881e0131 - std::rt::lang_start::{{closure}}::h73390e42b169ecf3
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/rt.rs:63:18
  32:     0x559a8992e72a - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::ha9408abe89f69dc4
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/ops/function.rs:259:13
  33:     0x559a8992e72a - std::panicking::try::do_call::h5b0cc9e9102acb65
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:401:40
  34:     0x559a8992e72a - std::panicking::try::hddc7f8229138b3ba
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:365:19
  35:     0x559a8992e72a - std::panic::catch_unwind::hfa401ff8bab2986e
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panic.rs:434:14
  36:     0x559a8992e72a - std::rt::lang_start_internal::{{closure}}::h8163422320d11405
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/rt.rs:45:48
  37:     0x559a8992e72a - std::panicking::try::do_call::hc742cc7bb4f0fb20
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:401:40
  38:     0x559a8992e72a - std::panicking::try::ha37d8d2dd1acf7d3
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:365:19
  39:     0x559a8992e72a - std::panic::catch_unwind::h8a5381d5ecf437bc
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panic.rs:434:14
  40:     0x559a8992e72a - std::rt::lang_start_internal::h7e2cee8c90d4a4d3
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/rt.rs:45:20
  41:     0x559a881e0100 - std::rt::lang_start::h17f9313942e84037
                               at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/rt.rs:62:5
  42:     0x559a88227dec - main
  43:     0x7fa1f72470b3 - __libc_start_main
  44:     0x559a88181fee - _start
  45:                0x0 - <unknown>
john@user-desktop:~/projects/sl/SL-test-viewer/slscenetester/target/debug$ 



```

**Comment by CatCode79** [comment]
I checked more carefully and am adequately convinced that the message
"EGL says it can present to the window but not natively [... and so on]"
is just a warning, the line is triggered because wgpu tries to use the backed Opengl Es but failing to configure it and ends up choosing the Vulkan backend.

Instead the index out of bounds crash is triggered [by this line](https://github.com/BVE-Reborn/rend3/blob/e6d19773bdc4754a566f6d803f06314fa59329aa/rend3/src/resources/material.rs#L422)
unfortunately i don't know rend3 and I got lost right away

**Comment by John-Nagle** [comment]
Oh, OK, if it ends up using the Vulkan back-end, that's what it should be doing anyway. I'm on Ubuntu Linux 20.04 LTS, where Vulkan is the standard graphics system.


Are you able to reproduce the bounds crash? 

It looks like somehow the registry table got out of sync with the material table. But no idea how that happened. It should not be possible to do that from outside Rend3, because handles, from which indices are obtained, are opaque reference-counted types.

**Comment by cwfitzgerald** [comment]
The EGL issues are harmless though annoying. If I remember correctly we can't actually tell what the non-native gpu is from the context of GL, so if we're on NV at all, we disable presentation. There might be some way of making this work better so as to not detect this machine.

As for the actual error, something is causing the material index map to get messed up such that it is reading one past the last element instead of the last element. This kinda smells like a bug with the use of swap_remove, though nothing jumped out at me immediately. I did however find two unrelated bugs with the new material system in some edge cases.

I will try to craft a repro case using one of the examples and work from there. Hopefully I can get to fixing these bugs shortly.

I want to add tests for things like this so I can be sure I don't break the (rather delicate) bookkeeping code.

**Comment by John-Nagle** [comment]
More info: This subscript panic does not happen if I load 50 objects or less. If I load 51 objects there is a subscript panic. Repeatable. With 50 objects, the whole system comes up, dynamic texture loading works, etc.

**Comment by John-Nagle** [comment]
More info: It's not the number of objects added to the scene, it's certain specific objects of mine that make it crash. If I skip specific objects by number that cause a crash, I can get many more objects loaded. Will try to figure out what is special about objects which cause a crash.

**Comment by John-Nagle** [comment]
More info: it looks like there are cases where I create a material and then discover I don't need it because the associated mesh is not present. So the material is created and immediately dropped without ever being used. Could that be causing trouble?

**Comment by cwfitzgerald** [comment]
This is amazing, thank you. Yeah the delete is very likely causing the problem, and also explains why it's trying to read exactly one past the end. Should make it quite ready to repro.

**Comment by John-Nagle** [comment]
More progress. Reordered my code so that object loading never creates a material and then immediately drops it. Crash stopped and entire scene will load.

Now getting a crash at program termination, as objects, materials, etc. are being dropped.

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/object.rs:90:70
stack backtrace:
   0: rust_begin_unwind
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:50:5
   3: core::option::Option<T>::unwrap
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/option.rs:722:21
   4: rend3::resources::object::ObjectManager::ready::{{closure}}
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/object.rs:90:25
   5: rend3::util::registry::archetypical::ArchetypicalRegistry<K,V,HandleType>::remove_all_dead
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/util/registry/archetypical.rs:87:21
   6: rend3::resources::object::ObjectManager::ready
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/resources/object.rs:88:9
   7: rend3::renderer::render::render_loop
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/renderer/render.rs:104:5
   8: rend3::renderer::Renderer::render
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/rend3-0.2.0/src/renderer/mod.rs:391:9
   9: slscenetester::displaystate::{{closure}}
             at /home/john/projects/sl/SL-test-viewer/slscenetester/src/main.rs:683:40
  10: winit::platform_impl::platform::sticky_exit_callback
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/mod.rs:746:5
  11: winit::platform_impl::platform::x11::EventLoop<T>::run_return
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/x11/mod.rs:307:21
  12: winit::platform_impl::platform::x11::EventLoop<T>::run
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/x11/mod.rs:385:9
  13: winit::platform_impl::platform::EventLoop<T>::run
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/mod.rs:662:56
  14: winit::event_loop::EventLoop<T>::run
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/event_loop.rs:154:9
  15: slscenetester::displaystate
             at /home/john/projects/sl/SL-test-viewer/slscenetester/src/main.rs:512:5
  16: slscenetester::main
             at /home/john/projects/sl/SL-test-viewer/slscenetester/src/main.rs:799:5
  17: core::ops::function::FnOnce::call_once
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

So, still trouble with deletion, but deferred until exit.

(Other problems: a few meshes which used to work are being corrupted, some UVs are bad, and sometimes, it takes several seconds to render a frame. More on that later.)

**Comment by cwfitzgerald** [comment]
I unfortunately can't seem to reproduce this locally by modifying one of the examples. The issue on exit seems to be basically the same issue where there's a desync between the material and object manager. I've finished all the main rendergraph work, so this is next on the docket. Going to work backwards from the panic and try to figure out the cause of the desync.

**Comment by John-Nagle** [comment]
Which problem can't be reproduced?

**Comment by cwfitzgerald** [comment]
Either of the desync issues reported in this thread. Still unsure what's causing #239.

**Comment by John-Nagle** [comment]
Hm. Maybe add some debug-mode asserts which, perhaps slowly, check for consistency after each API call? The crash occurs in the refresh loop, not at update time, so we don't get to see which API call back at update time triggered the error. If this could be caught at the API call, which is probably when the problem is created, it would be easier to find the cause.

**Comment by cwfitzgerald** [comment]
Yeah, that's a good idea, I need to add debug options for better error messages too which I can roll into that.

I'm investigating this right now, and I just found two different fairly glaring bugs that could be sources of the desync. I'm going to fix those and then see where we are.

**Comment by cwfitzgerald** [comment]
Closed by #243 and published as rend3/rend3-pbr 0.2.1.
